### DISCUSSION TITLE
v3.1 roadmap

### DISCUSSION THREAD

**Discussion by dlech** [pybricks/support#286] [Ideas]
This is the place to discuss ideas for the next Pybricks release.

Based on user feedback, the most common requests for consideration are:
- Spike Prime/ Robot Inventor hub support
- Hub-to-hub communication
- Remote control (LEGO Powered Up Handset or generic Bluetooth game controller)
- Code completion

Related discussions:
- https://github.com/pybricks/support/issues/262
- https://github.com/pybricks/support/issues/173
- https://github.com/pybricks/support/issues/29

---

For future visitors interested in the status of v3.1 (after discussion has taken place)... we will be tracking progress at  https://github.com/orgs/pybricks/projects/10

**Comment by Kefs88** [comment]
Hello Laurens and David,

Let me start by thanking you for your great work. I got back into Lego and quickly discovered that pybricks provided the control I needed that Lego did not.

It all started with ev3, and I was relieved to find a plethora of examples and tutorials to get me started as I'm new to coding. My main accomplishment here was getting my PS4 controller to work with the brick.

After that, I hoped I would be able to do the same for the (increasing number) of technic hubs I have laying around.

My vote for the next update would most definitely be controler support for the technic hub.

A close second would be hub-to-hub communication (so that I can control more hubs with a controller).


Keep up the good work!

**Comment by Novakasa** [comment]
- I would definitely appreciate a way to remotely start a stored program, see #284.
- I also really like the proposal for a pollable timer in #258 

In general I need to use PC-Hub-communication a lot, so everything that makes this more stable would be appreciated.
- It would be nice if there would be a user-facing API for sending data to the hub in `pybricksdev` (Right now we seem to rely on Bleak API: #333)
- Currenly I need to code a message input buffer myself to parse more complex input than a single char. It would be nice if something like this could be built in into the firmware, as I suspect coding it in python is not very efficient. Something like a `PybricksHub.send_message(msg: str)` in `pybricksdev` and something like this on pybricks side (mimmicking the API proposed for #258):

```python
from from pybricks.tools import Message
from uselect import poll
p = poll()
message = Message()
p.register(message)
for obj, _ in p.poll():
        if obj is message:
            handle_message(message.str)
```
- This could be combined with enabling the module `ujson` or similar ( https://github.com/pybricks/support/issues/140#issuecomment-803323760 though no Idea how much space it requires) to enable passing structured data. Right now I am just sending the repr of my python objects around, which works quite nicely actually, but I do have to call `eval` on the receiving side, which is a bit uncomfortable for me (though I don't think the security implications of arbitrary code execution on a Pybricks hub are that severe)

**Comment by laurensvalk** [comment]
We've rearranged the release priorities a bit, so please view the following issues for details.

- V3.1: #407
- V3.2: #406 

This thread will be closed.
